pkg/transport/process/utils/processor_message.go

```diff
@@ -0,0 +1,141 @@
+	ResourceBlobFile = "resource-blob"
+)
+
+// WriteProcessorMessage writes a component descriptor, resource and resource blob as a processor
+// message (tar archive with fixed filenames for component descriptor, resource, and resource blob)
+// which can be consumed by processors.
+func WriteProcessorMessage(cd cdv2.ComponentDescriptor, res cdv2.Resource, resourceBlobReader io.Reader, w io.Writer) error {
+	tw := tar.NewWriter(w)
+	defer tw.Close()
+
+	marshaledCD, err := yaml.Marshal(cd)
+	if err != nil {
+		return fmt.Errorf("unable to marshal component descriptor: %w", err)
+	}
+
+	if err := utils.WriteFileToTARArchive(ComponentDescriptorFile, bytes.NewReader(marshaledCD), tw); err != nil {
+		return fmt.Errorf("unable to write %s: %w", ComponentDescriptorFile, err)
+	}
+
+	marshaledRes, err := yaml.Marshal(res)
+	if err != nil {
+		return fmt.Errorf("unable to marshal resource: %w", err)
+	}
+
+	if err := utils.WriteFileToTARArchive(ResourceFile, bytes.NewReader(marshaledRes), tw); err != nil {
+		return fmt.Errorf("unable to write %s: %w", ResourceFile, err)
+	}
+
+	if resourceBlobReader != nil {
+		if err := utils.WriteFileToTARArchive(ResourceBlobFile, resourceBlobReader, tw); err != nil {
+			return fmt.Errorf("unable to write %s: %w", ResourceBlobFile, err)
+		}
+	}
+
+	return nil
+}
+
+// ReadProcessorMessage reads the component descriptor, resource and resource blob from a processor message
+// (tar archive with fixed filenames for component descriptor, resource, and resource blob) which is
+// produced by processors. The resource blob reader can be nil. If a non-nil value is returned, it must
+// be closed by the caller.
+func ReadProcessorMessage(r io.Reader) (*cdv2.ComponentDescriptor, cdv2.Resource, io.ReadCloser, error) {
+	tr := tar.NewReader(r)
+
+	var cd *cdv2.ComponentDescriptor
+	var res cdv2.Resource
+	var f *os.File
+
+	for {
+		header, err := tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, cdv2.Resource{}, nil, fmt.Errorf("unable to read tar header: %w", err)
+		}
+
+		switch header.Name {
+		case ResourceFile:
+			if res, err = readResource(tr); err != nil {
+				return nil, cdv2.Resource{}, nil, fmt.Errorf("unable to read %s: %w", ResourceFile, err)
+			}
+		case ComponentDescriptorFile:
+			if cd, err = readComponentDescriptor(tr); err != nil {
+				return nil, cdv2.Resource{}, nil, fmt.Errorf("unable to read %s: %w", ComponentDescriptorFile, err)
+			}
+		case ResourceBlobFile:
+			if f, err = ioutil.TempFile("", ""); err != nil {
+				return nil, cdv2.Resource{}, nil, fmt.Errorf("unable to create tempfile: %w", err)
+			}
+			if _, err := io.Copy(f, tr); err != nil {
+				return nil, cdv2.Resource{}, nil, fmt.Errorf("unable to read %s: %w", ResourceBlobFile, err)
+			}
+		}
+	}
+
+	if f == nil {
+		return cd, res, nil, nil
+	}
+
+	if _, err := f.Seek(0, io.SeekStart); err != nil {
+		return nil, cdv2.Resource{}, nil, fmt.Errorf("unable to seek to beginning of file: %w", err)
```

Comment:
why not returning the cd in error case?

Reply:
if an err is returned, no of the other results should be used anyways